lib/src/component/ref_util.dart

```diff
@@ -212,3 +214,282 @@ UiFactory<TProps> Function(UiFactory<TProps>) forwardRef<TProps extends UiProps>
 
   return wrapWithForwardRef;
 }
+
+/// Automatically passes a [Ref] through a component to one of its children.
```

Comment:
Some notes on this doc comment:
- The first sentence in the description could be updated to state a little more clearly what it does
- I think it'd be helpful to link the React JS docs up front, since it contains more information and can help introduce consumers to this concept
- We should show LHS typing in these, since that's our recommendation when declaring top-level function components.
- I think the examples could be adjusted to more closely resemble real-world use-cases, and that we could list the more common ones up front.
    
    For example, I think a good portion of the time, `uiForwardRef` forwarding to a DOM component is going to use a custom props class, and not just directly use `DomProps`. I think we should lead with recommending custom props classes over wrapping them with `asForwardRefConfig`.

- I think the class component HOC example (which I know was copied from `forwardRef`) is  a little convoluted, and could be improved.
- I think all of the imports and react_dom.render parts of the boilerplate can be omitted, since they add clutter/length to the examples

With that in mind, these are my suggested changes which address these points:
```dart
/// Creates a function component capable of forwarding its ref to
/// a component it renders.
///
/// Learn more: <https://reactjs.org/docs/forwarding-refs.html>.
///
/// ### Example 1: Updating a function component to forward a ref:
///
/// _This example mirrors the JS example in the link above._
///
/// Consider a `FancyButton` function component that renders the native button DOM element:
/// ```dart
/// mixin FancyButtonProps on UiProps {}
///
/// UiFactory<FancyButtonProps> FancyButton = uiFunction(
///   (props) {
///     return (Dom.button()
///       ..className = 'FancyButton'
///     )(props.children);
///   },
///   $FancyButtonConfig, // ignore: undefined_identifier
/// );
/// ```
///
/// Normally, you can't set a ref on function components, but if you change
/// [uiFunction] to `uiForwardRef`, you'll get a new `ref` argument that will be
/// populate with any ref set on `FancyButton`. We This ref can then be
/// "forwarded" to some other component.
///
/// ```dart
/// mixin FancyButtonProps on UiProps {}
///
/// UiFactory<FancyButtonProps> FancyButton = uiForwardRef(
///   (props, ref) {
///     return (Dom.button()
///       ..ref = ref
///       ..className = 'FancyButton'
///     )(props.children);
///   },
///   $FancyButtonConfig, // ignore: undefined_identifier
/// );
///
/// usageExample() {
///   // You can now get a ref directly to the DOM button:
///   final ref = createRef<ButtonElement>();
///   return (FancyButton()..ref = ref)();
/// }
/// ```
///
/// ### Example 2: Creating a [higher-order component](https://reactjs.org/docs/higher-order-components.html)
/// that forwards its ref to the wrapped component.
///
/// ```dart
/// /// Wraps a component ([factoryToWrap]) in a new component that logs when rendered.
/// UiFactory<TProps> withLogging<TProps extends UiProps>(UiFactory<TProps> factoryToWrap) {
///   return uiForwardRef(
///     (props, ref) {
///       useEffect(() => 'the component rendered!');
///
///       return (factoryToWrap()
///         ..addAll(props)
///         ..ref = ref
///       )(props.children);
///     },
///     factoryToBeWrapped.asForwardRefConfig(
///       displayName: 'withLogging(${getDisplayName(factoryToWrap)})',
///     ),
///   );
/// }
///
/// UiFactory<FancyButtonProps> FancyButton = ...;
///
/// /// This can be used just like FancyButton, and setting a ref on it will work the same!
/// UiFactory<FancyButtonProps> FancyButtonWithLogging = withLogging(FancyButton);
/// ```
///
/// ### Example 3: Exposing inner refs in class components
///
/// While it's usually easier to pass `uiForwardRef`'s `ref` argument directly
/// to the desired component, sometimes it's nested within a class component
/// that can't forward refs itself, and it's non-trivial to update that class
/// component to a function component.
///
/// In this case, you can pass the ref to the class component as a custom prop,
/// and have the class component forward it to the right location.
///
/// Here, we'll make our class component factory private and only expose a
/// version wrapped in `uiForwardRef`:
///
/// ```dart
/// UiFactory<LogPropsProps> _Foo = _$_Foo; // ignore: undefined_identifier
/// mixin FooProps on UiProps {
///   // Private since we only use this to pass along the ref provided in
///   // uiForwardRef.
///   //
///   // Consumers will effectively be setting this when they set `ref` on the public factory.
///   dynamic _inputRef;
/// }
/// class FooComponent extends UiComponent2<FooProps> {
///   @override
///   render() {
///     return Dom.div()(
///       (Dom.input()
///         ..type = 'text'
///         ..ref = _inputRef
///       )();
///     )
///   }
/// }
///
/// // ---------- Wrapping a Class Component ----------
/// // Here you have two options:
/// //   - Option 1: Use the class component's factory as the UiFactoryConfig arg.
/// //   This needs to be done because the builder recognizes
/// //   `FooProps` as already consumed (by the class component).
/// //
/// //   - Option 2: Create a new props class. This just works around the issue
/// //   described for Option 1 because it is creating a new props class, but it
/// //   only needs to mix in the props mixins that the original props class used.
/// //
/// // Choosing between the options is likely circumstantial or preferential.
/// // The advantage to Option 1 is that if the class component has numerous mixins,
/// // it is much more concise to create the function component config.
/// // Option 2 has the benefit that it matches the declaration of standard
/// // function components (which `uiForwardRef` returns). Additionally, Option 2
/// // illustrates how one could add additional props to the wrapped function component.
///
/// //
/// // Option 1 Example
/// UiFactory<FooProps> Foo = uiForwardRef((props, ref) {
///     return (_Foo()
///       ..addProps(props)
///       .._inputRef = ref
///     )();
///   },
///   _Foo.asForwardRefConfig(),
/// );
///
/// //
/// // Option 2 Example:
///
/// // This is not necessary but is just in place to illustrate that more props
/// // can be specified and consumed.
/// mixin AnotherPropsMixin on UiProps {
///   String anExampleAdditionalProp;
/// }
///
/// class Foo2Props = UiProps with AnotherPropsMixin, FooProps;
/// final Foo2 = uiForwardRef<LogsPropsComponent2Props>((props, ref) {
///     print(props.anExampleAdditionalProp);
///
///     return (_Foo()
///       ..addProps(props)
///       .._inputRef = ref
///     )();
///   },
///   $Foo2Config, // ignore: undefined_identifier
/// );
```

Reply:
Nice, I really like it. Thanks for typing that up!